src/runtime/virtcontainers/nydusd.go

```diff
@@ -0,0 +1,424 @@
+package virtcontainers
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/kata-containers/kata-containers/src/runtime/pkg/katautils/katatrace"
+	"github.com/kata-containers/kata-containers/src/runtime/virtcontainers/utils/retry"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	infoEndpoint  = "/api/v1/daemon"
+	mountEndpoint = "/api/v1/mount"
+
+	defaultHttpClientTimeout = 30 * time.Second
+	contentType              = "application/json"
+
+	maxIdleConns          = 10
+	idleConnTimeout       = 10 * time.Second
+	dialTimout            = 5 * time.Second
+	keepAlive             = 5 * time.Second
+	expectContinueTimeout = 1 * time.Second
+
+	nydusRafsType          = "rafs"
+	nydusPassthroughfsType = "passthrough_fs"
+
+	sharedPathInGuest = "/containers"
+
+	shimNsPath = "/proc/self/ns/net"
+)
+
+var (
+	nydusdTracingTags = map[string]string{
+		"source":    "runtime",
+		"package":   "virtcontainers",
+		"subsystem": "nydusd",
+	}
+
+	errNydusdDaemonPathEmpty    = errors.New("nydusd daemon path is empty")
+	errNydusdSocketPathEmpty    = errors.New("nydusd socket path is empty")
+	errNydusdAPISocketPathEmpty = errors.New("nydusd api socket path is empty")
+	errNydusdSourcePathEmpty    = errors.New("nydusd source path is empty")
+	errNydusdSourceNotAvailable = errors.New("nydusd source path not available")
+	errNydusdNotSupport         = errors.New("nydusd only support qemu currently")
```

Comment:
As mentioned above, can you include the URL for removing the QEMU restriction in the error message?

Reply:
has done